core/src/smartcontracts/isi/query.rs

```diff
@@ -6,283 +6,329 @@ use eyre::Result;
 use iroha_data_model::{
     prelude::*,
     query::{
-        error::QueryExecutionFail as Error, predicate::PredicateBox, Pagination, QueryOutputBox,
-        Sorting,
+        error::QueryExecutionFail as Error, parameters::IterableQueryParams, IterableQueryBox,
+        IterableQueryOutputBatchBox, QueryRequest, QueryRequestWithAuthority, QueryResponse,
+        SingularQueryBox, SingularQueryOutputBox,
     },
 };
-use parity_scale_codec::{Decode, Encode};
 
 use crate::{
-    prelude::ValidQuery,
+    prelude::ValidSingularQuery,
     query::{
-        cursor::{Batch as _, Batched},
-        pagination::Paginate as _,
+        cursor::QueryBatchedErasedIterator, pagination::Paginate as _, store::LiveQueryStoreHandle,
     },
+    smartcontracts::{wasm, ValidIterableQuery},
     state::{StateReadOnly, WorldReadOnly},
 };
 
-/// Represents lazy evaluated query output
-pub trait Lazy {
-    /// Type of the lazy evaluated query output
-    type Lazy<'a>;
+/// Allows to generalize retrieving the metadata key for all the query output types
+pub trait SortableQueryOutput {
+    /// Get the sorting key for the output, from metadata
+    ///
+    /// If the type doesn't have metadata or metadata key doesn't exist - return None
+    fn get_metadata_sorting_key(&self, key: &Name) -> Option<JsonString>;
 }
 
-/// Lazily evaluated equivalent of [`Query::Output`]
-pub enum LazyQueryOutput<'a> {
-    /// Concrete computed [`Query::Output`]
-    QueryOutput(QueryOutputBox),
-    /// Iterator over a set of [`Query::Output`]s
-    Iter(Box<dyn Iterator<Item = QueryOutputBox> + 'a>),
+impl SortableQueryOutput for Account {
```

Comment:
No, because of this limitation: https://github.com/rust-lang/rfcs/issues/2758.

Probably would've worked if this were not cross-crate (the type definitions are from data model, while the impl is in core), but alas